prow/config/agent.go

```diff
@@ -40,70 +45,186 @@ type Agent struct {
+					}
+				}
+			case err := <-w.Errors:
+				errFunc(err, fmt.Sprintf("received fsnotify error for directory %s", path))
+			}
+		}
+	}, nil
 }
 
-// Start will begin polling the config file at the path. If the first load
-// fails, Start will return the error and abort. Future load failures will log
-// the failure message but continue attempting to load.
-func (ca *Agent) Start(prowConfig, jobConfig string, additionals ...func(*Config) error) error {
-	lastModTime, err := lastConfigModTime(prowConfig, jobConfig)
+// GetFileWatcher returns a function that watches the specified file(s), running the "eventFunc" whenever an event for the file(s) occurs
+// and the "errFunc" whenever an error is encountered
+func GetFileWatcher(eventFunc func() error, errFunc func(error, string), files ...string) (func(ctx context.Context), error) {
+	w, err := fsnotify.NewWatcher()
 	if err != nil {
-		lastModTime = time.Time{}
+		return nil, err
 	}
-	c, err := Load(prowConfig, jobConfig, additionals...)
+	for _, file := range files {
+		err = w.Add(file)
+		if err != nil {
+			return nil, err
+		}
+	}
+	logrus.Debugf("Watching files: %v", files)
+	return func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				if err := w.Close(); err != nil {
+					errFunc(err, fmt.Sprintf("failed to close fsnotify watcher for files: %v", files))
+				}
+				return
+			case <-w.Events:
+				err := eventFunc()
+				if err != nil {
+					errFunc(err, fmt.Sprintf("event function failed watching files: %v", files))
+				}
+			case err := <-w.Errors:
+				errFunc(err, fmt.Sprintf("received fsnotify error watching files: %v", files))
+			}
+		}
+	}, nil
+}
+
+func watchConfigs(ca *Agent, prowConfig, jobConfig string, additionals ...func(*Config) error) error {
+	eventFunc := func() error {
+		c, err := Load(prowConfig, jobConfig, additionals...)
+		if err != nil {
+			return err
+		}
+		ca.Set(c)
+		return nil
+	}
+	errFunc := func(err error, msg string) {
+		logrus.WithField("prowConfig", prowConfig).
+			WithField("jobConfig", jobConfig).
+			WithError(err).Error(msg)
+	}
+	// List of all watcher functions to run
+	var runFuncs []func(context.Context)
+	prowIsCMMounted, err := IsConfigMapMount(filepath.Dir(prowConfig))
 	if err != nil {
 		return err
 	}
-	ca.Set(c)
-	go func() {
-		// Rarely, if two changes happen in the same second, mtime will
-		// be the same for the second change, and an mtime-based check would
-		// fail. Reload periodically just in case.
-		skips := 0
-		for range time.Tick(1 * time.Second) {
-			if skips < 600 {
-				recentModTime, err := lastConfigModTime(prowConfig, jobConfig)
+	// TODO(AlexNPavel): allow empty jobConfig till fully migrate config to subdirs
+	jobIsCMMounted := false
+	if jobConfig != "" {
+		jobIsCMMounted, err = IsConfigMapMount(filepath.Dir(jobConfig))
+		if err != nil {
+			return err
+		}
+	}
+	if filepath.Dir(prowConfig) == filepath.Dir(jobConfig) && prowIsCMMounted {
```

Comment:
Wow, that's a lot of error handling mess that really obfuscates what gets watched how and when... Maybe something like this would improve the readability and reduce duplication?

```
var cms string[]
var notcms string[]
WHAT IS WATCHED WHEN LOGIC
for cm in cms {
   runFunc, err := GetCMMountWatcher(eventFunc, errFunc, cm)
		if err != nil {
			return err
		}
		runFuncs = append(runFuncs, runFunc)
}
for notcm in notcms {
   runFunc, err := GetFileWatcher(eventFunc, errFunc, notcm)
		if err != nil {
			return err
		}
		runFuncs = append(runFuncs, runFunc)
}
```

Reply:
Yeah, that was overly complicated. Simplified based on your suggestions.